File: Data Structures & Algorithms/two-integer-sum/submission-4.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        std::unordered_map<int, std::vector<std::size_t>> num_to_idx;

        for (std::size_t idx = 0; idx < nums.size(); ++idx) {
            num_to_idx[nums[idx]].push_back(idx);
        }

        for (const auto& p : num_to_idx) {
            auto it = num_to_idx.find(target - p.first);

            if (it != num_to_idx.end()) {
                if (it->second.size() == 1) {
                    return {std::min(p.second[0], it->second[0]), std::max(p.second[0], it->second[0])};
                }
                else {
                    return {std::min(it->second[0], it->second[1]), std::max(p.second[0], it->second[1])};
                }
            }
        }

        return {};
    }
};

File: Data Structures & Algorithms/binary-search/submission-4.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        s, e = 0, len(nums) - 1

        while s <= e:
            diff = (e - s) // 2
            mid = s + diff
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                s = mid + 1
            else:
                e = mid - 1
        
        return -1